File: src/firefly/js/visualize/PlotState.js
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
import {isEmpty, isArray} from 'lodash';
import {Band} from './Band.js';
import {convertBandStateToJSON, makeBandState, makeBandStateWithJson} from './BandState.js';
import Enum from 'enum';


/**
 * @typedef RotateType
 * The type of rotation
 * can be 'NORTH', 'ANGLE', 'UNROTATE'
 * @prop NORTH
 * @prop ANGLE
 * @prop UNROTATE
 * @type {Enum}
 * @public
 * @global
 * */
export const RotateType= new Enum(['NORTH', 'ANGLE', 'UNROTATE']);

/**
 * @typedef Operation
 * can be 'ROTATE', 'CROP', 'FLIP_Y'
 * @prop CROP
 * @prop FLIP_Y
 * @type {Enum}
 */
export const Operation= new Enum(['ROTATE', 'CROP', 'FLIP_Y']);


export class PlotState {

    /**
     * @summary Contains data about the state of a plot.
     * This object is never created directly if is always instantiated from the json sent from the server.
     * @prop {boolean} threeColor - is a three color plot
     * @public
     */
    constructor() {
        this.bandStateAry= [null,null,null];
        this.ctxStr=null;
        this.threeColor= false;
        this.ops= [];
    }


//======================================================================
//----------------------- Public Methods -------------------------------
//======================================================================

    /**
     * @summary returns the first used band. It is possible that this method will return null.  You should always check.
     * @return {Band} the first name used.
     * @public
     */
    firstBand() {
        const bandAry= this.getBands();
        return bandAry && bandAry[0];
    }

    /**
     * @summary Get an array of used band.  It is possible that this routine will return a array of length 0
     * @return {Array} the bands in use
     * @public
     */
    getBands() {
        if (!this.usedBands) {
            this.usedBands= [];
            if (this.threeColor) {
                if (this.get(Band.RED).plotRequest)   this.usedBands.push(Band.RED);
                if (this.get(Band.GREEN).plotRequest) this.usedBands.push(Band.GREEN);
                if (this.get(Band.BLUE).plotRequest)  this.usedBands.push(Band.BLUE);
            }
            else {
                this.usedBands.push(Band.NO_BAND);
            }
        }
        return this.usedBands;
    }

    /**
     * @param {String|Band} [band] the bad to test
     * @return {boolean}
     * @public
     */
    isBandUsed(band) {
        return this.getBands().indexOf(band)>-1;
    }

    /**
     *
     * @return {boolean}
     */
    isThreeColor() { return this.threeColor; }

    /**
     * @summary this method will make a copy of WebPlotRequest. Any changes to the WebPlotRequest object
     * after the set will not be reflected here.
     * @param {Band} [band] the band to get the request for, if not passed the used the primary band
     * @return {WebPlotRequest} the WebPlotRequest
     * @public
     */
    getWebPlotRequest(band) { return this.get(band || this.firstBand()).plotRequest; }
    /**
     * @summary if a cube, checkout how many images it contains
     * @param {Band} [band] the band check for, if not passed the used the primary band
     * @return {number} the WebPlotRequest
     * @public
     */
    getCubeCnt(band) { return this.get(band || this.firstBand()).cubeCnt; }


    getCubePlaneNumber(band) {
        return this.get(band || this.firstBand()).cubePlaneNumber;
    }

    /**
     * Get the range values for the plot.
     * @param {Band} [band] the band get range value for, parameter is unnecessary for non-three color plots
     * @return {RangeValues}
     */
    getRangeValues(band) { return this.get(band || this.firstBand()).rangeValues; }

    setRangeValues(band,rv) { this.get(band).rangeValues= rv;}

    /**
     * @param band
     * @return {string}
     */
    getWorkingFitsFileStr(band=undefined) { return this.get(band || this.firstBand()).workingFitsFileStr;}

    /**
     * @param band
     * @return {string}
     */
    getOriginalFitsFileStr(band) { return this.get(band || this.firstBand()).originalFitsFileStr; }


    getUploadFileName(band) { return this.get(band || this.firstBand()).uploadFileNameStr; }

    /**
     *
     * @param {Operation} op
     */
    hasOperation(op) {
        let newOp;
        if (op.key) {
            newOp= op;
        }
        else if (typeof op.toString === 'function') {
            newOp= Operation.get(op.toString());
        }
        else if (typeof op === 'string') {
            newOp= Operation.get(op);
        }
        return newOp ? this.ops.indexOf(newOp)>-1  : false;
    }

    equals(obj) {
        return (obj instanceof PlotState) ? this.toJson()===obj.toJson() : false;
    }


    /**
     * @param {object} band
     * @return {BandState}
     */
    get(band) {
        let idx;
        if (band.value) {
            idx= band.value;
        }
        else if (typeof band === 'number') {
            idx= band;
        }
        else {
            const b= Band.get(band.toString());
            idx= b ? b.value : Band.NO_BAND.value;
        }

        if (!this.bandStateAry[idx]) this.bandStateAry[idx]= makeBandState();
        return this.bandStateAry[idx];
    }

    toJson(includeDirectAccessData= true) {
        return JSON.stringify(PlotState.convertToJSON(this, includeDirectAccessData));
    }

    copy() {
        return PlotState.makePlotStateWithJson(PlotState.convertToJSON(this,true));
    }

    static makePlotState() {
        return new PlotState();
    }

    static parse(jsonStr) {
        return PlotState.makePlotStateWithJson(JSON.parse(jsonStr));
    }

    /**
     * Create a plot state object from the json version of the object.
     * @param psJson
     * @param [overridePlotRequest] - if defined and not three color use this one instead of the on the the json object
     * @param [overrideRV] - if defined and not three color use this one instead of the on the the json object
     * @return {PlotState}
     */
    static makePlotStateWithJson(psJson, overridePlotRequest, overrideRV) {
        if (!psJson) return;
        const state= PlotState.makePlotState();
        // if not include used defaulted values
        state.threeColor= Boolean(psJson.threeColor);
        state.ops= psJson.ops?.map( (op) => Operation.get(op) ) ?? [];

        const {bandStateAry}= psJson;
        const ovPR= !state.threeColor ? overridePlotRequest : undefined;
        const ovRV= !state.threeColor ? overrideRV : undefined;

        if (isArray(bandStateAry)) {
            state.bandStateAry= bandStateAry.map( (bJ) => makeBandStateWithJson(bJ, ovPR,ovRV));
        }
        else {
            state.bandStateAry= [makeBandStateWithJson(bandStateAry, ovPR,ovRV)];
        }

        state.ctxStr=psJson.ctxStr;

        return state;
    }

    /**
     * @summary convert his PlotState to something can be used with JSON.stringify
     * @param {PlotState} s
     * @param {boolean} includeDirectAccessData include the includeDirectAccessData object
     */
    static convertToJSON(s, includeDirectAccessData= true) {
        if (!s) return undefined;
        const json= {};
        json.ctxStr=s.ctxStr;
        json.colorTableId= 0;

        if (!isEmpty(s.ops)) json.ops= s.ops.map( (op) => op.key );
        if (s.threeColor) json.threeColor= true;


        json.bandStateAry= s.bandStateAry.map( (bJ) => convertBandStateToJSON(bJ,includeDirectAccessData));
        return json;
    }

}

export function makePlotStateShimForHiPS(wpRequest) {
    const plotState= PlotState.makePlotState();
    const bandState= makeBandState();
    bandState.plotRequest= wpRequest;
    plotState.bandStateAry= [bandState,null,null];
    return plotState;
}


export default PlotState;

File: src/firefly/java/edu/caltech/ipac/firefly/visualize/PlotState.java
/*
 * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
 */
package edu.caltech.ipac.firefly.visualize;

import edu.caltech.ipac.util.ComparisonUtil;
import edu.caltech.ipac.visualize.plot.RangeValues;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Trey Roby
 */
public class PlotState {

    public enum Operation {CROP}
    public static final int MAX_BANDS= 3;


    // must be three, one for RED, GREEN, BLUE -
    // NO_BAND uses 0 like RED
    // NO_BAND and 3 color are mutually exclusive so there is no conflict
    private final BandState[] bandStateAry= new BandState[MAX_BANDS];


    private String ctxStr;
    private boolean threeColor = false;
    private List<Operation> ops = new ArrayList<>(1);

//======================================================================
//----------------------- Private Constructors -------------------------
//======================================================================

    public PlotState() { }

//======================================================================
//----------------------- Public Methods -------------------------------
//======================================================================

    public BandState[] getBandStateAry() { return bandStateAry; }

    public void setBandStateAry(BandState[] bsAry) {
        for(int i=0; (i<MAX_BANDS); i++) {
            if (i<bsAry.length) {
                bandStateAry[i]= bsAry[i];
            }
        }
    }

    public Band firstBand() {
        Band[] bandAry= getBands();
        return (bandAry!=null && bandAry.length>0) ? getBands()[0] : null;
    }

    Band[] NO_BAND_ARY= new Band[] {Band.NO_BAND};

    public Band[] getBands() {
        if (threeColor) {
            List<Band> bands= new ArrayList<>(3);
            if (get(Band.RED).hasRequest())   bands.add(Band.RED);
            if (get(Band.GREEN).hasRequest()) bands.add(Band.GREEN);
            if (get(Band.BLUE).hasRequest())  bands.add(Band.BLUE);
            return bands.toArray(new Band[0]);
        }
        else {
            return NO_BAND_ARY;
        }
    }

    public boolean isBandUsed(Band band) {
        Band[] bandAry= getBands();
        boolean retval= false;
        for(Band b : bandAry) {
            if (b==band) {
                retval= true;
                break;
            }
        }
        return retval;
    }

    public  String getContextString() { return ctxStr; }
    public  void setContextString(String ctxStr) { this.ctxStr = ctxStr; }

    public void setThreeColor(boolean threeColor) { this.threeColor = threeColor; }
    public boolean isThreeColor() { return threeColor; }

    /**
     * this method will make a copy of WebPlotRequest. Any changes to the WebPlotRequest object
     * after the set will not be reflected here.
     * @param plotRequests copy this request
     * @param band the band to set the request for
     */
    public void setWebPlotRequest(WebPlotRequest plotRequests, Band band) {
        setWebPlotRequest(plotRequests,band,true);
    }

    /**
     * this method will make a copy of WebPlotRequest. Any changes to the WebPlotRequest object
     * after the set will not be reflected here.
     * @param plotRequests copy this request
     * @param band the band to set the request for
     */
    public void setWebPlotRequest(WebPlotRequest plotRequests, Band band, boolean initStretch) {
        get(band).setWebPlotRequest(plotRequests);
        if (initStretch) initColorStretch(plotRequests,band);
    }

    /**
     * this method will make a copy of WebPlotRequest. Any changes to the WebPlotRequest object
     * after the set will not be reflected here.
     * @param band the band to get the request for
     * @return the WebPlotRequest
     */
    public WebPlotRequest getWebPlotRequest(Band band) {
        if (band==null) band= firstBand();
        return get(band).getWebPlotRequest();
    }

    /**
     * this method will make a copy of WebPlotRequest. Any changes to the WebPlotRequest object
     * after the set will not be reflected here.
     * @return the WebPlotRequest
     */
    public WebPlotRequest getPrimaryRequest() { return get(firstBand()).getWebPlotRequest(); }



    /**
     * this method will make a copy of the primary WebPlotRequest. Any changes to the WebPlotRequest object
     * after the set will not be reflected here.
     * @return the WebPlotRequest
     */
    public WebPlotRequest getWebPlotRequest() { return get(firstBand()).getWebPlotRequest(); }

    public boolean isMultiImageFile(Band band) {
        if (band==null) band= firstBand();
        return get(band).isMultiImageFile();
    }

    public boolean isMultiImageFile() { return get(firstBand()).isMultiImageFile(); }
    public void setMultiImageFile(boolean multiImageFile, Band band) { get(band).setMultiImageFile(multiImageFile); }

    public int getCubeCnt(Band band) {
        if (band==null) band= firstBand();
        return get(band).getCubeCnt();
    }
    public int getCubeCnt() { return get(firstBand()).getCubeCnt(); }
    public void setCubeCnt(int cubeCnt, Band band) { get(band).setCubeCnt(cubeCnt); }

    public int getCubePlaneNumber(Band band) {
        if (band==null) band= firstBand();
        return get(band).getCubePlaneNumber();
    }

    public int getCubePlaneNumber() { return get(firstBand()).getCubePlaneNumber(); }
    public void setCubePlaneNumber(int cubeIdx, Band band) { get(band).setCubePlaneNumber(cubeIdx); }


    public void setRangeValues(RangeValues rv, Band band) { get(band).setRangeValues(rv); }
    public RangeValues getRangeValues(Band band) {
        if (band==null) band= firstBand();
        return get(band).getRangeValues();
    }

    public RangeValues getRangeValues() { return get(firstBand()).getRangeValues(); }

    public BandState.FileAndHeaderInfo getFileAndHeaderInfo(Band band) {
        return get(band).getFileAndHeaderInfo();
    }

    public String getWorkingFitsFileStr(Band band) { return band!=null ? get(band).getWorkingFitsFileStr() : null; }
    public void setWorkingFitsFileStr(String fileStr, Band band) { get(band).setWorkingFitsFileStr(fileStr); }

    public String getOriginalFitsFileStr(Band band) { return band!=null ? get(band).getOriginalFitsFileStr() : null; }
    public void setOriginalFitsFileStr(String fileStr, Band band) { get(band).setOriginalFitsFileStr(fileStr); }

    public String getUploadFileName(Band band) { return band!=null ? get(band).getUploadedFileName() : null; }
    public void setUploadFileName(String fileStr, Band band) { get(band).setUploadedFileName(fileStr); }

    public void setImageIdx(int idx, Band band) { get(band).setImageIdx(idx);}
    public int getImageIdx(Band band) { return get(band).getImageIdx(); }

    public void setOriginalImageIdx(int idx, Band band) { get(band).setOriginalImageIdx(idx); }
    public int getOriginalImageIdx(Band band) { return get(band).getOriginalImageIdx(); }

    public void addOperation(Operation op) {if (!ops.contains(op)) ops.add(op); }
    public boolean hasOperation(Operation op) {return ops.contains(op); }
    public List<Operation> getOperations() { return ops;}

    public String toPrettyString() {
        return "ctxStr: " + ctxStr + ", 3 color: " + threeColor;
    }

    public String toString() { return toPrettyString(); }

    public PlotState makeCopy() {
        PlotState p= new PlotState();
        p.ctxStr= this.ctxStr;
        p.threeColor= this.threeColor;
        p.ops = this.ops;
        for(int i= 0; (i< bandStateAry.length); i++) {
            if (this.bandStateAry[i]!=null) {
                p.bandStateAry[i]= this.bandStateAry[i].makeCopy();
            }
        }
        return p;
    }

    public boolean equals(Object o) {
        if (o==this) return true;
        if (!(o instanceof PlotState ps)) return false;
        return ComparisonUtil.equals(bandStateAry, ps.bandStateAry) &&
                ComparisonUtil.equals(ctxStr, ps.ctxStr) &&
                threeColor == ps.threeColor;
    }

    public BandState get(Band band) {
        int idx= band.getIdx();
        if (bandStateAry[idx]==null) bandStateAry[idx]= new BandState();
        return bandStateAry[idx];
    }

// =====================================================================
// -------------------- private Methods --------------------------------
// =====================================================================

    private void initColorStretch(WebPlotRequest request, Band band) {
        if (request!=null) {
            if (request.containsParam(WebPlotRequest.INIT_RANGE_VALUES)) {
                String rvStr= request.getParam(WebPlotRequest.INIT_RANGE_VALUES);
                RangeValues rv= RangeValues.parse(rvStr);
                if (rv!=null) get(band).setRangeValues(rv);
            }
        }

    }
}

